File: contexts/typescript/services/log_store.ts
// Log store — reads the JSON-lines log the app writes. Your real log
// aggregation (Loki/CloudWatch/ELK/...) goes here instead. ESSENCE §4A.
import { existsSync, readFileSync } from "node:fs";

export interface LogLine {
  ts: number;
  level: string;
  text: string;
}

export class FileLogStore {
  path: string;
  constructor(p: string) {
    this.path = p;
  }

  fetch(since = 0, until = Infinity): LogLine[] {
    if (!existsSync(this.path)) return [];
    return readFileSync(this.path, "utf8")
      .split("\n")
      .filter((l) => l.trim().length > 0)
      .map((l) => JSON.parse(l) as LogLine)
      .filter((r) => r.ts >= since && r.ts <= until);
  }
}

File: contexts/python/services/log_store.py
"""Log store — reads the JSON-lines log the app writes. Your real log
aggregation (Loki/CloudWatch/ELK/...) goes here instead. ESSENCE §4A."""
import json
import os


class FileLogStore:
    def __init__(self, path: str) -> None:
        self.path = path

    def fetch(self, since: float = 0.0, until: float | None = None) -> list[dict]:
        hi = float("inf") if until is None else until
        out: list[dict] = []
        if not os.path.exists(self.path):
            return out
        with open(self.path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                rec = json.loads(line)
                if since <= rec.get("ts", 0.0) <= hi:
                    out.append(rec)
        return out
